Data/Pool.hs

```diff
@@ -403,6 +437,16 @@ destroyResource Pool{..} LocalPool{..} resource = do
    atomically (modifyTVar_ inUse (subtract 1))
 {-# INLINABLE destroyResource #-}
 
+giveResource :: Pool a -> a -> IO ()
+giveResource pool@Pool{..} resource = do
+  local@LocalPool{..} <- getLocalPool pool
+  join $ atomically $ do
+    currentMaximum <- readTVar maxResources
+    currentlyUsed <- readTVar inUse
+    if currentlyUsed > currentMaximum
+      then return $ destroyResource pool local resource
+      else return $ putResource local resource
```

Comment:
IIUC, if `currentlyUsed > currentMaximum` we try to destroy the resource because we have already used the max number of connections and it will not be possible to re-use the resource?

Reply:
Yes, by destroying the resource, we immediately reduce the pool size, which in the `currentlyUsed > currentMaximum` scenario is what we want.